marker_driver_rustc/src/conversion/marker/expr.rs

```diff
@@ -153,6 +153,20 @@ impl<'ast, 'tcx> MarkerConverterInner<'ast, 'tcx> {
             hir::ExprKind::Field(operand, field) => {
                 ExprKind::Field(self.alloc(FieldExpr::new(data, self.to_expr(operand), self.to_ident(*field))))
             },
+            hir::ExprKind::If(scrutinee, then, els) => ExprKind::If(self.alloc(IfExpr::new(
+                data,
+                self.to_expr(scrutinee),
+                self.to_expr(then),
+                els.map(|els| self.to_expr(els)),
+            ))),
+            hir::ExprKind::Let(lets) => self.to_let_expr(lets),
+            hir::ExprKind::Match(scrutinee, arms, hir::MatchSource::Normal) => {
+                ExprKind::Match(self.alloc(MatchExpr::new(data, self.to_expr(scrutinee), self.to_match_arms(arms))))
+            },
+            // `DropTemps` is an rustc internal construct to tweak the drop
+            // order during HIR lowering. Marker can for now ignore this and
+            // convert the inner expression directly
+            hir::ExprKind::DropTemps(inner) => self.to_expr(inner),
```

Comment:
Do we actually want to pass these through? I don't see what value would the developer of lint crates would get from rustc internal constructs, triggering lints on things user didn't write seems odd.

And even if a developer needs to interact with these, passing them unwrapped (as normal expressions) would make it impossible to distinguish from the user written expressions.

Reply:
I don't think we want to. These could be relevant for users trying to analyze when objects are dropped, as they restrict the scope of values. With this, a user could see right away, that the condition of an if expression triggers a drop for created values. Without it users need to read the documentation to see this. However, as you said, it's weird if the API receives expressions that weren't created by the user. I also believe that rustc is the only driver using this representation. Adding it could also be a maintenance burden later down the line. So, in total, I would just ignore this and its implications.

Here is a playground example that has a `DropTemps` expression, and I think the AST is simpler without it: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=859f8ca4c65cd558fa5cce0c7870faf1

Reply:
Oh, I misunderstood the purpose of this variant I think. Thanks for the quick explanation and a playground example. Yes I think this change is totally valid then, you're free to ignore my comment :)